executor/new_executor_write.go

```diff
@@ -0,0 +1,264 @@
+	return nil
+}
+
+// Next implements Executor Next interface.
+func (e *NewUpdateExec) Next() (*Row, error) {
+	if !e.fetched {
+		err := e.fetchRows()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		e.fetched = true
+	}
+
+	columns, err := getNewUpdateColumns(e.OrderedList)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if e.cursor >= len(e.rows) {
+		return nil, nil
+	}
+	if e.updatedRowKeys == nil {
+		e.updatedRowKeys = make(map[table.Table]map[int64]struct{})
+	}
+	row := e.rows[e.cursor]
+	newData := e.newRowsData[e.cursor]
+	for _, entry := range row.RowKeys {
+		if entry == nil { // outer join
+			continue
+		}
+		tbl := entry.Tbl
+		if e.updatedRowKeys[tbl] == nil {
+			e.updatedRowKeys[tbl] = make(map[int64]struct{})
+		}
+		offset := e.getTableOffset(*entry)
+		handle := entry.Handle
+		oldData := row.Data[offset : offset+len(tbl.WritableCols())]
+		newTableData := newData[offset : offset+len(tbl.WritableCols())]
+		_, ok := e.updatedRowKeys[tbl][handle]
+		if ok {
+			// Each matching row is updated once, even if it matches the conditions multiple times.
+			continue
+		}
+		// Update row
+		err1 := newUpdateRecord(e.ctx, handle, oldData, newTableData, columns, tbl, offset, false)
+		if err1 != nil {
+			return nil, errors.Trace(err1)
+		}
+		e.updatedRowKeys[tbl][handle] = struct{}{}
+	}
+	e.cursor++
+	return &Row{}, nil
+}
+
+func getNewUpdateColumns(assignList []*expression.Assignment) (map[int]*expression.Assignment, error) {
+	m := make(map[int]*expression.Assignment, len(assignList))
+	for i, v := range assignList {
+		m[i] = v
+	}
+	return m, nil
+}
+
+func (e *NewUpdateExec) fetchRows() error {
+	for {
+		row, err := e.SelectExec.Next()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if row == nil {
+			return nil
+		}
+		data := make([]types.Datum, len(e.SelectExec.Schema()))
+		newData := make([]types.Datum, len(e.SelectExec.Schema()))
+		for i, s := range e.SelectExec.Schema() {
+			data[i], err = s.Eval(row.Data, e.ctx)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			newData[i] = data[i]
+			if e.OrderedList[i] != nil {
+				val, err := e.OrderedList[i].Expr.Eval(row.Data, e.ctx)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				newData[i] = val
+			}
+		}
+		row.Data = data
+		e.rows = append(e.rows, row)
+		e.newRowsData = append(e.newRowsData, newData)
+	}
+}
+
+func (e *NewUpdateExec) getTableOffset(entry RowKeyEntry) int {
+	t := entry.Tbl
+	var tblName string
+	if entry.TableAsName == nil || len(entry.TableAsName.L) == 0 {
+		tblName = t.Meta().Name.L
+	} else {
+		tblName = entry.TableAsName.L
+	}
+	schema := e.SelectExec.Schema()
+	for i := 0; i < len(schema); i++ {
+		s := schema[i]
+		if s.TblName.L == tblName {
+			return i
+		}
+	}
+	return 0
+}
+
+func newUpdateRecord(ctx context.Context, h int64, oldData, newData []types.Datum, updateColumns map[int]*expression.Assignment, t table.Table, offset int, onDuplicateUpdate bool) error {
```

Comment:
This function is not needed, we can reuse old `updateRecord` method, just change `updateColumns `map[int]*ast.Assignment` to `map[int]bool`, as the value of the map just used to check `nil`, not actually used.